sync/statesync/code_queue.go

```diff
@@ -0,0 +1,260 @@
+func NewCodeQueue(db ethdb.Database, quit <-chan struct{}, opts ...CodeQueueOption) (*CodeQueue, error) {
+	// Apply defaults then options.
+	o := codeQueueOptions{
+		capacity: defaultQueueCapacity,
+	}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	q := &CodeQueue{
+		db:   db,
+		out:  make(chan common.Hash, o.capacity),
+		quit: quit,
+	}
+
+	// Close the output channel on early shutdown to unblock consumers.
+	go func() {
+		<-q.quit
+		// If not already closed, mark as quit and close channel.
+		q.closed.markQuitAndClose(q.out)
+	}()
+
+	// Always initialize eagerly.
+	if err := q.init(); err != nil {
+		return nil, err
+	}
+
+	return q, nil
+}
+
+// CodeHashes returns the receive-only channel of code hashes to consume.
+func (q *CodeQueue) CodeHashes() <-chan common.Hash {
+	return q.out
+}
+
+// AddCode implements [syncpkg.CodeRequestQueue] by persisting and enqueueing new hashes.
+// Persists idempotent "to-fetch" markers for all inputs and enqueues them as-is.
+// Returns errAddCodeAfterFinalize after a clean finalize and errFailedToAddCodeHashesToQueue on early quit.
+func (q *CodeQueue) AddCode(codeHashes []common.Hash) error {
+	// If the queue has been closed, reject new work.
+	switch q.closed.get() {
+	case closeStateFinalized:
+		return errAddCodeAfterFinalize
+	case closeStateQuit:
+		return errFailedToAddCodeHashesToQueue
+	}
+	return q.enqueue(codeHashes)
+}
+
+// init enqueues any persisted code markers found on disk.
+func (q *CodeQueue) init() error {
+	// Recover any persisted code markers and enqueue them.
+	// Note: dbCodeHashes are already present as "to-fetch" markers. addCode will
+	// re-persist them, which is a trivial redundancy that happens only on resume
+	// (e.g., after restart). We accept this to keep the code simple.
+	dbCodeHashes, err := recoverUnfetchedCodeHashes(q.db)
+	if err != nil {
+		return fmt.Errorf("unable to recover previous sync state: %w", err)
+	}
+	if err := q.enqueue(dbCodeHashes); err != nil {
+		return fmt.Errorf("unable to resume previous sync: %w", err)
+	}
+
+	return nil
+}
+
+// enqueue persists markers for the provided hashes and pushes them onto the output channel.
+// No deduplication is performed. The consumer is responsible for dedupe and retry.
+func (q *CodeQueue) enqueue(codeHashes []common.Hash) error {
+	if len(codeHashes) == 0 {
+		return nil
+	}
+
+	batch := q.db.NewBatch()
+	// Persist all input hashes as to-fetch markers. Consumer will dedupe and skip
+	// already-present code. Persisting all enables consumer-side retry.
+	// Note: markers are keyed by code hash, so repeated persists overwrite the same
+	// key rather than growing DB usage. The consumer deletes the marker after
+	// fulfilling the request (or when it detects code is already present).
+	selected := make([]common.Hash, 0, len(codeHashes))
+	for _, codeHash := range codeHashes {
+		customrawdb.AddCodeToFetch(batch, codeHash)
+		selected = append(selected, codeHash)
+	}
```

Comment:
```suggestion
	for _, codeHash := range codeHashes {
		customrawdb.AddCodeToFetch(batch, codeHash)
	}
```

With consumer-side deduplication, `selected` is an unnecessary clone of `codeHashes`. You'll also need to change the `range` below.